app/src/main/java/com/google/samples/apps/sunflower/adapters/ComposeRecyclerViewUtils.kt

```diff
@@ -0,0 +1,62 @@
+/*
+ * Copyright 2021 Google LLC
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.google.samples.apps.sunflower.adapters
+
+import androidx.compose.runtime.Composable
+import androidx.compose.ui.platform.ComposeView
+import androidx.compose.ui.platform.ViewCompositionStrategy
+import androidx.recyclerview.widget.DiffUtil
+import androidx.recyclerview.widget.ListAdapter
+import androidx.recyclerview.widget.RecyclerView
+
+abstract class ComposeListAdapter<T, VH : ComposeViewHolder<T>>(
+    diffCallback: DiffUtil.ItemCallback<T>
+) : ListAdapter<T, VH>(diffCallback) {
+
+    override fun onViewRecycled(holder: VH) {
+        holder.composeView.disposeComposition()
+        super.onViewRecycled(holder)
+    }
+}
+
+abstract class ComposeRecyclerViewAdapter<VH : ComposeViewHolder<*>> : RecyclerView.Adapter<VH>() {
+
+    override fun onViewRecycled(holder: VH) {
+        holder.composeView.disposeComposition()
+        super.onViewRecycled(holder)
+    }
+}
+
+abstract class ComposeViewHolder<T>(
+    val composeView: ComposeView
```

Comment:
Should we force extenders to construct the `ComposeView` or construct it internally, or even as a default? Might consumers want to configure the `ComposeView` they pass in?

Reply:
We cannot construct it internally because `RecyclerView.ViewHolder` takes the `View` in its constructor. 

We could have it as a default but then it'd also require a `Context` to be passed in as a parameter. That might be even more confusing:

```
abstract class ComposeViewHolder<T>(
    context: Context,
    val composeView: ComposeView = ComposeView(context)
) : RecyclerView.ViewHolder(composeView) {
```

Not sure how often you want to configure the `ComposeView` passed in, maybe you want to change the layout params of it?